src/mlpack/core/data/split_data.hpp

```diff
@@ -368,6 +341,241 @@ Split(const arma::Mat<T>& input,
                          std::move(testData));
 }
 
+/**
+ * Given an input dataset and labels, split into a training set and test set.
+ * Example usage below.  This overload places the split dataset into the four
+ * output parameters given (trainData, testData, trainLabel, and testLabel).
+ *
+ * The input dataset must be of type arma::field. It should have the shape -
+ * (n_rows = 1, n_cols = Number of samples, n_slices = 1)
+ *
+ * NOTE: Here FieldType could be arma::field<arma::mat> or arma::field<arma::vec>
+ *
+ * @code
+ * arma::field<arma::mat> input = loadData();
+ * arma::field<arma::vec> label = loadLabel();
+ * arma::field<arma::mat> trainData;
+ * arma::field<arma::mat> testData;
+ * arma::field<arma::vec> trainLabel;
+ * arma::field<arma::vec> testLabel;
+ * math::RandomSeed(100); // Set the seed if you like.
+ *
+ * // Split the dataset into a training and test set, with 30% of the data being
+ * // held out for the test set.
+ * Split(input, label, trainData,
+ *                testData, trainLabel, testLabel, 0.3);
+ * @endcode
+ *
+ * @param input Input dataset to split.
+ * @param inputLabel Input labels to split.
+ * @param trainData FieldType to store training data into.
+ * @param testData FieldType test data into.
+ * @param trainLabel Field vector to store training labels into.
+ * @param testLabel Field vector to store test labels into.
+ * @param testRatio Percentage of dataset to use for test set (between 0 and 1).
+ * @param shuffleData If true, the sample order is shuffled; otherwise, each
+ *       sample is visited in linear order. (Default true.)
+ */
+template <typename FieldType,
+          typename = std::enable_if_t<
+              arma::is_Col<typename FieldType::object_type>::value ||
+              arma::is_Mat_only<typename FieldType::object_type>::value>>
+void Split(FieldType& input,
+           arma::field<arma::vec>& inputLabel,
+           FieldType& trainData,
+           arma::field<arma::vec>& trainLabels,
+           FieldType& testData,
+           arma::field<arma::vec>& testLabels,
+           const double testRatio,
+           const bool shuffleData = true)
+{
+  const size_t testSize = static_cast<size_t>(input.n_cols * testRatio);
+  const size_t trainSize = input.n_cols - testSize;
+
+  trainData.set_size(1, trainSize);
+  testData.set_size(1, testSize);
+
+  arma::uvec order = arma::linspace<arma::uvec>(0, input.n_cols - 1,
+      input.n_cols);
+
+  if (shuffleData)
+    order = arma::shuffle(order);
+
+  if (trainSize > 0)
+  {
+    trainLabels.set_size(1, trainSize);
+
+    for (size_t i = 0; i < trainSize; ++i)
+      trainData[i] = input(0, order(i));
+
+    for (size_t i = 0; i < trainSize; ++i)
+      trainLabels(0, i) = inputLabel(0, order(i));
+  }
```

Comment:
I haven't checked every line, but this looks really similar to the other overload.  Could we merge them, perhaps?

Reply:
The two overloads are almost similar. They have only one difference i.e. one can be used when we have labels and the other can be used when we do not have labels.

Reply:
Oh, sorry, I mean the other overload where we don't take an `arma::field`.  So, e.g., we could have one combined overload for when we have labels, and one combined overload for when we don't, regardless of whether we have an `arma::field` as input.  Let me know what you think. :+1:

Reply:
Hii @rcurtin , I analyzed the code and I found this repetitive pattern for all splitting overloads.
We are doing the following logic for multiple overloads.
```
if (shuffle_data == True):
    order <-- array of shuffled indexes
    train_data <-- elements of input from shuffled indices
    test_data <-- remaining elements of input
else:
    train_data <-- starting few elements of input
    test_data <-- remaining elements of input
```

in some places, these assignments to `train_data` and `test_data` are done using armadillo functions and at some places using a for loop. But the logic is the same everywhere. So, I was thinking if rather than trying to merge the overloads, we can create a `SplitHelper` function that can perform this logic. This will hugely decrease the unnecessary lines of repetitive code.

Kindly let me know your take on this.

Reply:
That sounds like a good idea to me.  I suspect that we might find that there is a way to write all the code in one way, but perhaps a first step towards seeing if we can do that is to refactor into a `SplitHelper()` like you suggested. :+1: